View transition old snapshots incorrectly get scaled to the border-box size.

mattwoodrow opened the pull request:
#### 34e14bf3631972e46abcee10b0b72f0339e7a493
<pre>
View transition old snapshots incorrectly get scaled to the border-box size.
<a href="https://bugs.webkit.org/show_bug.cgi?id=269954">https://bugs.webkit.org/show_bug.cgi?id=269954</a>
&lt;<a href="https://rdar.apple.com/123478049">rdar://123478049</a>&gt;

Reviewed by Tim Nguyen.

Drawing of the old snapshot rescales the image to the border-box size, even if
we captured a larger area.

We shouldn&apos;t apply scaling, just align the origin of the border-box in the snapshot
to the border-box origin of the pseudo element.

This records the overflow rect for which the snapshot is captured, and uses it
to adjust the overflow rect of the pseudo element and align the painting
position of the snapshot.

As a slight cleanup, merges the two createRendererIfNeeded lambdas into a single
static function.

* LayoutTests/TestExpectations:
* LayoutTests/platform/glib/TestExpectations:
* Source/WebCore/dom/ViewTransition.cpp:
(WebCore::snapshotNodeVisualOverflowClippedToViewport):
(WebCore::ViewTransition::captureOldState):
* Source/WebCore/dom/ViewTransition.h:
* Source/WebCore/page/FrameSnapshotting.cpp:
(WebCore::snapshotFrameRectWithClip):
* Source/WebCore/page/FrameSnapshotting.h:
* Source/WebCore/rendering/RenderBoxModelObject.cpp:
(WebCore::RenderBoxModelObject::requiresLayer const):
* Source/WebCore/rendering/RenderElement.cpp:
(WebCore::RenderElement::capturedInViewTransition const):
(WebCore::RenderElement::hasViewTransitionName const):
* Source/WebCore/rendering/RenderElement.h:
* Source/WebCore/rendering/RenderInline.cpp:
(WebCore::RenderInline::requiresLayer const):
* Source/WebCore/rendering/RenderLayer.cpp:
(WebCore::canCreateStackingContext):
(WebCore::RenderLayer::shouldBeCSSStackingContext const):
(WebCore::RenderLayer::computeCanBeBackdropRoot const):
* Source/WebCore/rendering/RenderTableRow.cpp:
(WebCore::RenderTableRow::requiresLayer const):
* Source/WebCore/rendering/RenderViewTransitionCapture.cpp:
(WebCore::RenderViewTransitionCapture::setImage):
(WebCore::RenderViewTransitionCapture::paintReplaced):
(WebCore::RenderViewTransitionCapture::layout):
* Source/WebCore/rendering/RenderViewTransitionCapture.h:
* Source/WebCore/rendering/updating/RenderTreeUpdaterViewTransition.cpp:
(WebCore::createRendererIfNeeded):
(WebCore::RenderTreeUpdater::ViewTransition::buildPseudoElementGroup):
(WebCore::RenderTreeUpdater::ViewTransition::updatePseudoElementGroup):

Canonical link: <a href="https://commits.webkit.org/275227@main">https://commits.webkit.org/275227@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/17d825e9d94472f43c416805b7ca9991e0742ca7

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/41255 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/20269 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/43633 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/43819 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/37347 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/43562 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/23332 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/17599 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/34117 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/41829 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/17187 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/35534 "Passed tests") | [  ~~🧪 api-wpe~~](https://ews-build.webkit.org/#/builders/41/builds/14774 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/42/builds/14918 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/10/builds/36536 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/45145 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/37434 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/36854 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/40592 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/16070 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/13178 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/38978 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/17689 "Built successfully") | | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/9248 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/17741 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/17333 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

nt1m commented on Source/WebCore/dom/ViewTransition.cpp:
```diff
@@ -260,6 +262,32 @@ static ExceptionOr<void> checkDuplicateViewTransitionName(const AtomString& name
+    frame.view()->setNodeToDraw(&node);
+
+    ASSERT(node.renderer()->hasLayer());
+    RenderLayerModelObject* layerRenderer = downcast<RenderLayerModelObject>(node.renderer());
```
```suggestion
    auto* layerRenderer = downcast<RenderLayerModelObject>(node.renderer());
```

(Also does this need a smart pointer?)

nt1m commented on Source/WebCore/dom/ViewTransition.cpp:
```diff
@@ -260,6 +262,32 @@ static ExceptionOr<void> checkDuplicateViewTransitionName(const AtomString& name
+        return nullptr;
+
+    OptionSet<PaintBehavior> paintBehavior = frame.view()->paintBehavior();
+    Color backgroundColor = frame.view()->baseBackgroundColor();
```
```suggestion
    auto paintBehavior = frame.view()->paintBehavior();
    auto backgroundColor = frame.view()->baseBackgroundColor();
```

nt1m commented on Source/WebCore/dom/ViewTransition.cpp:
```diff
@@ -260,6 +262,32 @@ static ExceptionOr<void> checkDuplicateViewTransitionName(const AtomString& name
+
+    IntRect paintRect = layerRenderer->layer()->absoluteBoundingBox();
+    paintRect.intersect(snappedIntRect(frame.view()->layoutViewportRect()));
+    RefPtr<ImageBuffer> result = snapshotFrameRect(frame, paintRect, WTFMove(options));
```
```suggestion
    RefPtr result = snapshotFrameRect(frame, paintRect, WTFMove(options));
```

nt1m commented on Source/WebCore/rendering/RenderElement.cpp:
```diff
@@ -2066,12 +2066,20 @@ bool RenderElement::hasSelfPaintingLayer() const
     return !!document().activeViewTransition();
 }
 
+bool RenderElement::hasViewTransitionProperty() const
```
```suggestion
bool RenderElement::hasViewTransitionName() const
```

but also why the split?

nt1m commented on Source/WebCore/rendering/RenderViewTransitionCapture.cpp:
```diff
@@ -36,38 +36,35 @@ RenderViewTransitionCapture::RenderViewTransitionCapture(Type type, Document& do
-    if (paintInfo.phase == PaintPhase::Foreground)
-        page().addRelevantRepaintedObject(*this, intersection(replacedContentRect, contentBoxRect));
+    IntPoint pos = snappedIntRect(replacedContentRect).location();
+    pos.moveBy(roundedIntPoint(m_overflowRect.location()));
```
nit: no abbreviations in WebKit:  `position`

alanbaradlay commented on Source/WebCore/rendering/RenderElement.cpp:
```diff
@@ -2066,12 +2066,20 @@ bool RenderElement::hasSelfPaintingLayer() const
+
+    return true;
+}
+
```
why not `return style().viewTransitionName();` ?

nt1m commented on Source/WebCore/dom/ViewTransition.cpp:
```diff
@@ -260,6 +262,34 @@ static ExceptionOr<void> checkDuplicateViewTransitionName(const AtomString& name
     return { };
 }
 
+static RefPtr<ImageBuffer> snapshotNodeVisualOverflowClippedToViewport(LocalFrame& frame, Node& node, SnapshotOptions&& options, LayoutRect& oldOverflowRect)
```
I don't have a strong opinion but this could also go into FrameSnapshotting.cpp, also fine in ViewTransition.cpp to me. Up to you.

nt1m reviewed (approved):
Please make sure you rebaseline GLIB and update the changelog before landing.

nt1m commented on Source/WebCore/dom/ViewTransition.cpp:
```diff
@@ -260,6 +262,34 @@ static ExceptionOr<void> checkDuplicateViewTransitionName(const AtomString& name
+    frame.view()->setBaseBackgroundColor(Color::transparentBlack);
+    frame.view()->setNodeToDraw(&node);
+
+    ASSERT(node.renderer()->hasLayer());
```
This demo hits the assert FYI: https://codepen.io/argyleink/full/abPOmgj

mattwoodrow replied:
Fixed, was missing `RenderTableRow::requiresLayer`.

mattwoodrow committed:
View transition old snapshots incorrectly get scaled to the border-box size.
https://bugs.webkit.org/show_bug.cgi?id=269954
<rdar://123478049>

Reviewed by Tim Nguyen.

Drawing of the old snapshot rescales the image to the border-box size, even if
we captured a larger area.

We shouldn't apply scaling, just align the origin of the border-box in the snapshot
to the border-box origin of the pseudo element.

This records the overflow rect for which the snapshot is captured, and uses it
to adjust the overflow rect of the pseudo element and align the painting
position of the snapshot.

As a slight cleanup, merges the two createRendererIfNeeded lambdas into a single
static function.

* LayoutTests/TestExpectations:
* LayoutTests/platform/glib/TestExpectations:
* Source/WebCore/dom/ViewTransition.cpp:
(WebCore::snapshotNodeVisualOverflowClippedToViewport):
(WebCore::ViewTransition::captureOldState):
* Source/WebCore/dom/ViewTransition.h:
* Source/WebCore/page/FrameSnapshotting.cpp:
(WebCore::snapshotFrameRectWithClip):
* Source/WebCore/page/FrameSnapshotting.h:
* Source/WebCore/rendering/RenderBoxModelObject.cpp:
(WebCore::RenderBoxModelObject::requiresLayer const):
* Source/WebCore/rendering/RenderElement.cpp:
(WebCore::RenderElement::capturedInViewTransition const):
(WebCore::RenderElement::hasViewTransitionName const):
* Source/WebCore/rendering/RenderElement.h:
* Source/WebCore/rendering/RenderInline.cpp:
(WebCore::RenderInline::requiresLayer const):
* Source/WebCore/rendering/RenderLayer.cpp:
(WebCore::canCreateStackingContext):
(WebCore::RenderLayer::shouldBeCSSStackingContext const):
(WebCore::RenderLayer::computeCanBeBackdropRoot const):
* Source/WebCore/rendering/RenderTableRow.cpp:
(WebCore::RenderTableRow::requiresLayer const):
* Source/WebCore/rendering/RenderViewTransitionCapture.cpp:
(WebCore::RenderViewTransitionCapture::setImage):
(WebCore::RenderViewTransitionCapture::paintReplaced):
(WebCore::RenderViewTransitionCapture::layout):
* Source/WebCore/rendering/RenderViewTransitionCapture.h:
* Source/WebCore/rendering/updating/RenderTreeUpdaterViewTransition.cpp:
(WebCore::createRendererIfNeeded):
(WebCore::RenderTreeUpdater::ViewTransition::buildPseudoElementGroup):
(WebCore::RenderTreeUpdater::ViewTransition::updatePseudoElementGroup):

Canonical link: https://commits.webkit.org/275227@main
